text/0000-deref-move.md

```diff
@@ -0,0 +1,325 @@
+- Feature Name: `deref_move`
+- Start Date: 2018-05-05
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Add a new `DerefMove` trait that allows consuming a smart pointer to move its
+contents, as in `let x = *p;`.
+
+For more background and discussion, see #178 and #997.
+
+# Motivation
+[motivation]: #motivation
+
+Currently, two smart pointer traits provide access to values contained in smart
+pointers: `Deref` and `DerefMut`. These traits allow overloading of the
+dereferencing (unary `*`) operator, and by extension participate in autoderef in
+a number of places, such as method calls and index expressions.
+
+These two traits, however, only dereference references to produce references to
+the contained value. They do not offer the option to dereference by value,
+consuming the smart pointer to produce the owned value inside. As a special
+case, `Box<T>` has compiler support allowing it to be dereferenced (explicitly
+or implicitly) to produce the inner `T`. Because this is special-cased in the
+compiler, this functionality is not available for other smart pointer types.
+
+Other smart pointers may support this consuming operation, but it must be done
+explicitly due to the lack of language support. For instance, in the stdlib, the
+following methods both consume a smart pointer to produce the inner value:
+
+* `ManuallyDrop<T>::into_inner()`
+* `binary_heap::PeakMut<'a, T>::pop()`
+* `Cow<'a, B>::into_owned()`
+
+I have not made any attempt to survey into the larger ecosystem for more
+examples, but there are probably many more.
+
+Note that removing `FnBox` is explicitly *not* part of the motivation for this
+RFC. It has been suggested in a few places, such as in discussion of
+rust-lang/rust#28796, that `DerefMove` would allow `Box<FnOnce()>` to be called
+without needing the `FnBox` workaround. This is not the case, however: the inner
+value is dynamically-sized and thus cannot be moved onto the stack. Indeed, one
+can call a `Box<T>` where `T` is a concrete type implementing `FnOnce()`.
+`DerefMove` would allow `Box<FnOnce()>` to continue not working, just without a
+special case in the compiler to allow it to do so.
+
+This will not remove all of `Box`'s special cases from the compiler, but
+represents progress towards that goal.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The following trait is defined in `std::ops`:
+
+```rust
+trait DerefMove : DerefMut {
```

Comment:
The rationale is below, but I had not considered this use case. The ergonomic drawbacks may be worth it if this use case is worth supporting.

Reply:
The behaviour of ignoring `DerefMove` when `Target : Copy` for `Deref` would be complicated if the `Target` types were different; that would possibly create scenarios where a generic smart pointer would fail for a non-`Copy` `Target` because `DerefMove` is ignored but `Deref` has a different `Target`.

This would also complicate autoderef and in particular its relationship with overload resolution, probably unbearably so. I think it would be so hard to use a type with different `Deref` and `DerefMove` `Target`s that I think that, if we decided to make the two independently implementable, the compiler must enforce that the `Target`s are identical.

Reply:
Hmm, I suppose something like `struct TaggedPtr(usize);` is a case where it could plausibly be `DerefMove` but not `Deref` -- the move version is cheap, just masking away the extra bits in the alignment, and conceptually reasonable, since it's still a pointer -- but it can't be `Deref` because it doesn't actually have the real pointer to borrow.

That would need that "split a trait into two" RFC, I think, to introduce `DerefTarget` or something that just provides the associated type.  But it might also be too much of a stretch to bother supporting even if we could...

Reply:
Small nit: this trait requires a `Sized` super trait and `where Self::Target: Sized` bound.

Reply:
@scottmcm The `TaggedPtr` case could still be `Deref`, I think, since `usize` is `Copy`, as long as the target was the pointed-to value, not the pointer itself.  It would require unsafe code to cast the masked `usize` to an actual pointer with a lifetime, but that would require unsafe code anyway--it would still (presumably) be safe.

Reply:
@pythonesque Good point!  I forget that pointers get to cheat 🙂